pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java

```diff
@@ -397,11 +401,54 @@ public BrokerService(PulsarService pulsar, EventLoopGroup eventLoopGroup) throws
                         BrokerService.class.getClassLoader());
 
         this.brokerEntryPayloadProcessors = BrokerEntryMetadataUtils.loadInterceptors(pulsar.getConfiguration()
-                        .getBrokerEntryPayloadProcessors(), BrokerService.class.getClassLoader());
+                .getBrokerEntryPayloadProcessors(), BrokerService.class.getClassLoader());
 
         this.bundlesQuotas = new BundlesQuotas(pulsar.getLocalMetadataStore());
     }
 
+    public void addTopicEventListener(TopicEventsListener... listeners) {
+        Objects.requireNonNull(listeners);
+        for (TopicEventsListener listener : listeners) {
+            if (listener != null) {
+                topicEventListeners.add(listener);
+            }
+        }
+        getTopics().keys().forEach(topic ->
+                notifyTopicListeners(listeners, topic, TopicEvent.AFTER_TOPIC_LOAD));
+    }
+
+    private void notifyTopicListeners(String topic, TopicEvent event) {
+        for (TopicEventsListener listener : topicEventListeners) {
+            try {
+                listener.handleEvent(topic, event);
+            } catch (Throwable t) {
+                log.error("TopicEventsListener exception while handling {} for topic {}", event, topic, t);
+            }
+        }
+    }
+    private <T> CompletableFuture<T> notifyTopicListeners(CompletableFuture<T> future, String topic,
+                                                          TopicEvent successEvent,
+                                                          TopicEvent failureEvent) {
+        future.whenComplete((r, ex) -> notifyTopicListeners(topic,
+                ex == null ? successEvent : failureEvent));
```

Comment:
We'll want to be clear about which thread these callbacks are run on to prevent deadlock scenarios.

Reply:
I thought about this. changing .whenComple to .whenCompleteAsync is easy but if I use regular thread pool I can get events firing out of order, maintaining another ordered thread pool for the sake of feature that won't be used by majority of users since this is needed for KOP) seems unnecessary. 
I suggest we let whoever implements the listener take care of not blocking the thread.

Reply:
That makes sense to me. As long as the design is documented, I think it's fair to say that we want to be conservative about adding extra threads and tread pools.